### PR TITLE
ARROW-10085: [C++] Fix S3 region resolution on Windows

### PR DESCRIPTION
**Pull request by pitrou**
Avoid trying to following redirects when resolving a region, as that can fail
(at least on Windows with certain AWS SDK versions).

Also skip a test that's failing because of a bug in the AWS SDK on Windows.

### CODE REVIEW

File: cpp/src/arrow/filesystem/s3fs.cc [hunk 1]
```diff
@@ -459,6 +459,24 @@ class S3Client : public Aws::S3::S3Client {
...
+    typename std::enable_if<HaveLegacyFollowRedirects, Config*>::type config) {
+  config->followRedirects = false;
+}
+
+template <typename Config = Aws::Client::ClientConfiguration>
+void DisableRedirects(
+    typename std::enable_if<!HaveLegacyFollowRedirects, Config*>::type config) {
+  config->followRedirects = decltype(config->followRedirects)::NEVER;
+}
+
```
Comment (pitrou) [inline cpp/src/arrow/filesystem/s3fs.cc:479]: I couldn't think of anything better. @bkietz

**Comment by pitrou** [comment]
CI failures are unrelated.

File: cpp/src/arrow/filesystem/s3fs.cc (hunk 1 above)
Comment (bkietz) [inline cpp/src/arrow/filesystem/s3fs.cc:479]: Maybe a little better?
```suggestion
// In AWS SDK < 1.8, Aws::Client::ClientConfiguration::followRedirects is a bool.
void DisableRedirectsImpl(bool* followRedirects) { *followRedirects = false; }

// In AWS SDK >= 1.8, it's a Aws::Client::FollowRedirectsPolicy scoped enum.
template <typename PolicyEnum, PolicyEnum Never = PolicyEnum::NEVER>
void DisableRedirectsImpl(PolicyEnum* followRedirects) { *followRedirects = Never; }

void DisableRedirects(Config* c) {
  DisableRedirectsImpl(&c->followRedirects);
}
```

Comment (pitrou) [reply cpp/src/arrow/filesystem/s3fs.cc:479]: I hadn't thought of that. WIll try it out, thank you.

Comment (pitrou) [reply cpp/src/arrow/filesystem/s3fs.cc:479]: Unfortunately:
```
../src/arrow/filesystem/s3fs.cc:463:6: error: unused function 'DisableRedirectsImpl' [-Werror,-Wunused-function]
void DisableRedirectsImpl(bool* followRedirects) { *followRedirects = false; }
     ^
```

Comment (pitrou) [reply cpp/src/arrow/filesystem/s3fs.cc:479]: `if constexpr` would have been nice...

**Comment by pitrou** [comment]
cc @nealrichardson

**Comment by nealrichardson** [comment]
@github-actions crossbow submit homebrew-r-autobrew

File: cpp/src/arrow/filesystem/s3fs.cc (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/filesystem/s3fs.cc:479]:
```diff
-void DisableRedirectsImpl(bool* followRedirects) { *followRedirects = false; }
+template <bool Never = false>
+void DisableRedirectsImpl(bool* followRedirects) { *followRedirects = Never; }
```

Comment (pitrou) [reply cpp/src/arrow/filesystem/s3fs.cc]: Seems to work!

**Review by bkietz — APPROVED** [review]

**Comment by bkietz** [comment]
Passing Travis: https://travis-ci.org/github/pitrou/arrow/builds/730048965

**Pull request closed without merging**